pkg/object/pipeline/spec.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pipeline
+
+import (
+	"fmt"
+
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/supervisor"
+	"github.com/megaease/easegress/pkg/util/yamltool"
+	"github.com/megaease/easegress/pkg/v"
+)
+
+type (
+	// Spec describes the Pipeline.
+	Spec struct {
+		Name     string                   `yaml:"-" jsonschema:"-"`
+		Protocol context.Protocol         `yaml:"protocol" jsonschema:"required"`
+		Flow     []Flow                   `yaml:"flow" jsonschema:"omitempty"`
+		Filters  []map[string]interface{} `yaml:"filters" jsonschema:"required"`
+	}
+
+	// Flow controls the flow of pipeline.
+	Flow struct {
+		Filter string `yaml:"filter" jsonschema:"required,format=urlname"`
+	}
+
+	// FilterSpec is the universal spec for all filters.
+	FilterSpec struct {
+		super *supervisor.Supervisor
+
+		rawSpec    map[string]interface{}
+		yamlConfig string
+		meta       *FilterMetaSpec
+		filterSpec interface{}
+		rootFilter Filter
```

Comment:
the name 'rootFilter' is also confusing, I think this field could be removed by refactor the code.